dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,202 @@
+
+        gdrive://1r3UbnmS5B4-7YZPZmyqJuCxLVps1mASC
+
+        (get it https://drive.google.com/drive/folders/{here})
+
+    - Dataset named "my-dataset" in the hidden application folder:
+
+        gdrive://appDataFolder/my-dataset
+
+        (this one wouldn't be visible through Google Drive web UI and
+         couldn't be shared)
+    """
+
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"google-auth-oauthlib": google_auth_oauthlib}
+    PARAM_CHECKSUM = "md5Checksum"
+    SPACE_DRIVE = "drive"
+    SCOPE_DRIVE = "https://www.googleapis.com/auth/drive"
+    SPACE_APPDATA = "appDataFolder"
+    SCOPE_APPDATA = "https://www.googleapis.com/auth/drive.appdata"
+    DEFAULT_OAUTH_ID = "default"
+
+    # Default credential is needed to show the string of "Data Version
+    # Control" in OAuth dialog application name and icon in authorized
+    # applications list in Google account security settings. Also, the
+    # quota usage is limited by the application defined by client_id.
+    # The good practice would be to suggest the user to create their
+    # own application credentials.
+    DEFAULT_CREDENTIALPATH = os.path.join(
+        os.path.dirname(__file__), "google-dvc-client-id.json"
+    )
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.root = self.path_info.netloc.lower()
+        if self.root == self.SPACE_APPDATA.lower():
+            default_scopes = self.SCOPE_APPDATA
+            space = self.SPACE_APPDATA
+        else:
+            default_scopes = self.SCOPE_DRIVE
+            space = self.SPACE_DRIVE
+        if Config.SECTION_GDRIVE_CREDENTIALPATH not in config:
+            shared_token_warning()
+            credentialpath = config.get(
+                Config.SECTION_GDRIVE_CREDENTIALPATH,
+                self.DEFAULT_CREDENTIALPATH,
+            )
+        scopes = config.get(Config.SECTION_GDRIVE_SCOPES, default_scopes)
+        # scopes should be a list and it is space-delimited in all
+        # configs, and `.split()` also works for a single-element list
+        scopes = scopes.split()
+
+        core_config = self.repo.config.config[Config.SECTION_CORE]
+        oauth2_flow_runner = core_config.get(
+            Config.SECTION_CORE_OAUTH2_FLOW_RUNNER, "console"
+        )
+
+        self.client = GDriveClient(
+            space,
+            config.get(Config.SECTION_GDRIVE_OAUTH_ID, self.DEFAULT_OAUTH_ID),
+            credentialpath,
+            scopes,
+            oauth2_flow_runner,
+        )
+
+    def get_file_checksum(self, path_info):
+        metadata = self.client.get_metadata(path_info, fields=["md5Checksum"])
+        return metadata["md5Checksum"]
+
+    def exists(self, path_info):
+        return self.client.exists(path_info)
+
+    def batch_exists(self, path_infos, callback):
```

Comment:
@efiop do we need to update anything to support threading here - I mean status, etc? you are changing something with @pared as far as I understand.

Reply:
Yes, batch_exists will be no longer needed after https://github.com/iterative/dvc/pull/2375 . As to threads in general, as long as self.client is thread-safe(looks like it is, but maybe @ei-grad could confirm/deny that) we will be fine.